src/roswire/definitions/package.py

```diff
@@ -90,7 +90,22 @@ def paths(shell: ShellProxy) -> List[str]:
         code, path_str, duration = shell.execute('echo "${ROS_PACKAGE_PATH}"')
         if code != 0:
             raise Exception("unexpected error when fetching ROS_PACKAGE_PATH")
-        paths: List[str] = path_str.strip().split(':')
+        package_paths: List[str] = path_str.strip().split(':')
+        paths: List[str] = []
+        for path in package_paths:
+            print("P: {}".format(path))
+            code, _, _ = shell.execute('test -f {}/package.xml'.format(path))
+            print("code: {}".format(code))
+            if code != 0:
+                paths.append(path)
+            else:
+                code, dirs_str, _ = shell.execute('ls -d {}/*'.format(path))
+                print("dirs_str: {}".format(dirs_str))
+                if code != 0:
+                    raise Exception("unexpected error when listing directories")
+                dirs: List[str] = path_str.strip().split('\n')
+                package_paths.extend(dirs)
```

Comment:
It shouldn't be possible to modify a list whilst iterating over it. Maybe switch to using a Queue instead?